diff --git a/WORKSPACE b/WORKSPACE
index 725ab8d9..d7652fed 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -366,5 +366,5 @@ http_archive(
     build_file = "@//:BUILD.minizip",
-    sha256 = "9a93b2b7dfdac77ceba5a558a580e74667dd6fede4585b91eefb60f03b72df23",
+    sha256 = "17e88863f3600672ab49182f217281b6fc4d3c762bde361935e436a95214d05c",
     strip_prefix = "zlib-1.3.1/contrib/minizip",
-    url = "https://zlib.net/fossils/zlib-1.3.1.tar.gz",
+    url = "https://github.com/madler/zlib/archive/refs/tags/v1.3.1.tar.gz",
 )
